crates/ide/src/folding_ranges.rs

```diff
@@ -146,8 +163,93 @@ pub(crate) fn folding_ranges(file: &SourceFile) -> Vec<Fold> {
     res
 }
 
-fn fold_kind(kind: SyntaxKind) -> Option<FoldKind> {
-    match kind {
+/// Builds a fold for the given syntax element.
+///
+/// This function creates a `Fold` object that represents a collapsible region in the code.
+/// If `collapsed_text` is enabled, it generates a preview text for certain fold kinds that
+/// shows a summarized version of the folded content.
+fn build_fold(element: &SyntaxElement, kind: FoldKind, collapsed_text: bool) -> Fold {
+    if !collapsed_text {
+        return Fold::new(element.text_range(), kind);
+    }
+
+    let fold_with_collapsed_text = match kind {
+        FoldKind::TailExpr => {
+            let expr = ast::Expr::cast(element.as_node().unwrap().clone()).unwrap();
```

Comment:
We should be able to get rid of this unwrap dance by making `FoldKind` contain the appropriate syntax element, that is
```rust
enum FoldKind {
    TailExpr(ast::Expr),
    Stmt(ast::Stmt),
}
```